### ISSUE TITLE
Elegant way of operating with nan values in MultiHeadAttention layer

### ISSUE THREAD

**Issue by markub3327** [keras-team/keras#16108]
Hello,

  I need to control the MultiHeadAttention layer by attention mask. I have nan values in input too and needed to be filtered out by an attention mask. Attention mask is generated ineffectively way by the nan values contained in the input tensor with `tf.matmul` operation.

What is the best way to generate an attention mask?
Please code a feature to work with nan values filtered by the attention mask.
It can be efficiently if I use SparseTensor and how to apply it on this issue?

**System information**
- TensorFlow version: 2.7.0

Based on: https://chat.stackoverflow.com/rooms/242271/discussion-between-bc-martin-kubovcik-and-alonetogether

## Code
```python
import tensorflow as tf
import numpy as np


x = tf.constant([[[        np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan],
        [        np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan],
        [        np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan],
        [        np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan],
        [        np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan],
        [        np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan],
        [        np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan],
        [        np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan,
                 np.nan,         np.nan,         np.nan,         np.nan],
        [ 0.7558127 ,  1.5447265 ,  1.6315602 , -0.19868968,
          0.08828261,  0.01711658, -1.8133892 ,  0.12930395,
          0.47128937,  0.08567389, -1.7158676 , -0.5843805 ],
        [-0.7664911 , -0.7145203 , -1.089696  ,  0.14649415,
          0.03585422,  0.9916008 ,  0.9384322 ,  0.34755042,
         -0.09592161,  0.76490027, -1.2517685 , -1.5740465 ]]])

print(x)

# generate mask
attention_mask = tf.cast(
    tf.math.reduce_all(
        tf.math.logical_not(tf.math.is_nan(x)), axis=-1
    ),
    dtype=tf.float32,
)
attention_mask = tf.matmul(
    attention_mask[..., tf.newaxis],
    attention_mask[..., tf.newaxis],
    transpose_b=True,
)

print(attention_mask)

l = tf.keras.layers.MultiHeadAttention(1, 32)
y = l(x, x, x, return_attention_scores=True, attention_mask=attention_mask)

print(y)
```

Thanks a lot for your time.

**Comment by chenmoneygithub** [comment]
@markub3327 Thanks for reporting the issue! 

I am not sure if I understand your question correctly, do you mean this step is slow?
```python
attention_mask = tf.matmul(
    attention_mask[..., tf.newaxis],
    attention_mask[..., tf.newaxis],
    transpose_b=True,
)
```
An alternative to this matrix multiplication is to use simple dimension cast, i.e.:
```
attention_mask = attention_mask[:, tf.newaxis, :]
l = tf.keras.layers.MultiHeadAttention(1, 32)
y = l(x, x, x, return_attention_scores=True, attention_mask=attention_mask)
```
Notice that it does not generate a symmetric attention matrix, but you can still filter out outputs you do not want using the original 2D mask.

**Comment by markub3327** [comment]
Yes, I think that.

I will try the simple dimension cast method. Yes, I see it you only add an axis and create a matrix with only one row, which can work as good as a big 2D table.

But what about NaN values, Will be filtered or a must wait to fix the issue and new release?

Thanks a lot.

**Comment by chenmoneygithub** [comment]
Usually models do not expect to take NaN values, so I would add an extra step to replace NaN values with 0, and filter out the output involving NaN using the mask.

**Comment by markub3327** [comment]
Thanks a lot for the explanation.

**Comment by chenmoneygithub** [comment]
np! I will close this issue, and feel free to reopen it if you have more questions, thanks!
